services/gitdiff/gitdiff.go

```diff
@@ -555,6 +283,180 @@ func DiffInlineWithHighlightCode(fileName, language, code string) DiffInline {
 	return DiffInline{EscapeStatus: status, Content: template.HTML(content)}
 }
 
+// HighlightCodeDiff is used to do diff with highlighted HTML code.
+// The HTML tags will be replaced by Unicode placeholders: "<span>{TEXT}</span>" => "\uE000{TEXT}\uE001"
+// These Unicode placeholders are friendly to the diff.
+// Then after diff, the placeholders in diff result will be recovered to the HTML tags.
+// It's guaranteed that the tags in final diff result are paired correctly.
+type HighlightCodeDiff struct {
+	placeholderBegin    rune
+	placeholderMaxCount int
+	placeholderIndex    int
+	placeholderTagMap   map[rune]string
+	tagPlaceholderMap   map[string]rune
+
+	lineWrapperTags []string
+}
+
+func NewHighlightCodeDiff() *HighlightCodeDiff {
+	return &HighlightCodeDiff{
+		placeholderBegin:    rune(0xE000), // Private Use Unicode: U+E000..U+F8FF, BMP(0), 6400
+		placeholderMaxCount: 6400,
+		placeholderTagMap:   map[rune]string{},
+		tagPlaceholderMap:   map[string]rune{},
```

Comment:
Perhaps it would be a good idea performance-wise to initialize the map with a capacity of `placeholderMaxCount` entries:
`make(map[rune]string{}, placeholderMaxCount)`
While it will waste memory in most cases, it certainly never needs to be resized (consistently), which will improve performance

Reply:
It would be too large. Usually there would only be a few placeholders.

And for small maps, the pre-allocation seems not improving too much performance, especially compared to other heavy logic, the pre-allocation here seems very trivial. We can benchmark it 😊

Reply:
What would you say is a typical count for small, medium, and large files respectively?
A guess is enough.

Reply:
Very small. These tags comes from Chroma's highlight styles.

And the code does diff line by line, every line only have a few styles.

So, you do not need to worry about the 6400 limit, it won't happen. I just put it there to make the logic complete.